src/VisualStudio/Core/Def/RoslynPackage.cs

```diff
@@ -82,6 +74,25 @@ internal class RoslynPackage : AbstractPackage
             _solutionEventMonitor = new SolutionEventMonitor(_workspace);
         }
 
+        private async Task InitializeFromBackgroundAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            FatalError.Handler = FailFast.OnFatalException;
+            FatalError.NonFatalHandler = WatsonReporter.Report;
+
+            // We also must set the FailFast handler for the compiler layer as well
+            var compilerAssembly = typeof(Compilation).Assembly;
+            var compilerFatalError = compilerAssembly.GetType("Microsoft.CodeAnalysis.FatalError", throwOnError: true);
+            var property = compilerFatalError.GetProperty(nameof(FatalError.Handler), BindingFlags.Static | BindingFlags.Public);
+            var compilerFailFast = compilerAssembly.GetType(typeof(FailFast).FullName, throwOnError: true);
+            var method = compilerFailFast.GetMethod(nameof(FailFast.OnFatalException), BindingFlags.Static | BindingFlags.NonPublic);
+            property.SetValue(null, Delegate.CreateDelegate(property.PropertyType, method));
+
+            _componentModel = (IComponentModel)await GetServiceAsync(typeof(SComponentModel)).ConfigureAwait(true);
```

Comment:
❓ Is `SComponentModel` _documented_ as free-threaded? If not, you need to switch to the main thread before this line.

Reply:
it is thread-safe. but also, ComponentModel is not COM, so there is no QI involved. GetServiceAsync can be called from BG for any service according to Ryan.

so. this should safe. I need to change ConfigureAwait(false) though.